Android/Samples/libraries/kml/src/main/java/com/microsoft/maps/kml/KMLParser.java

```diff
@@ -588,34 +692,91 @@ private void verifyElementSeen(@NonNull String tag, boolean hasSeenTag) throws K
     }
   }
 
+  private void verifyIdNotInMap(Map map, String id) throws KMLParseException {
+    if (map.containsKey(id)) {
+      throw new KMLParseException(
+          "Error at: " + mParser.getPositionDescription() + " ID " + id + " already seen.");
+    }
+  }
+
+  private void mergeStyles() {
+    for (StylesHolder inlineStyle : mMergeStyles.keySet()) {
+      StylesHolder sharedStyle = mMapStylesHolders.get(mMergeStyles.get(inlineStyle));
+      if (sharedStyle.getIconStyle().getImage() != null) {
+        inlineStyle.getIconStyle().setImage(sharedStyle.getIconStyle().getImage());
+      }
+      LineStyle sharedLineStyle = sharedStyle.getLineStyle();
+      LineStyle inlineLineStyle = inlineStyle.getLineStyle();
+      if (!inlineLineStyle.useWidth() && sharedLineStyle.useWidth()) {
+        inlineLineStyle.setWidth(sharedLineStyle.getWidth());
+        inlineLineStyle.setUseWidth(true);
+      }
+      if (!inlineLineStyle.useStrokeColor() && sharedLineStyle.useStrokeColor()) {
+        inlineLineStyle.setStrokeColor(sharedLineStyle.getStrokeColor());
+        inlineLineStyle.setUseStrokeColor(true);
+      }
+      PolyStyle sharedPolyStyle = sharedStyle.getPolyStyle();
+      PolyStyle inlinePolyStyle = inlineStyle.getPolyStyle();
+      if (!inlinePolyStyle.useFillTag() && sharedPolyStyle.useFillTag()) {
+        inlinePolyStyle.setShouldFill(sharedPolyStyle.shouldFill());
+        inlinePolyStyle.setUseFillTag(true);
+      }
+      if (!inlinePolyStyle.useOutlineTag() && sharedPolyStyle.useOutlineTag()) {
+        inlinePolyStyle.setShouldOutline(sharedPolyStyle.shouldOutline());
+        inlinePolyStyle.setUseOutlineTag(true);
+      }
+    }
+  }
+
   private void applyStyles() throws KMLParseException {
     for (String id : mMapElementStyles.keySet()) {
-      StylesHolder stylesHolder = mMapStylesHolders.get(id);
+      StylesHolder stylesHolder =
+          mMapStylesHolders.containsKey(id)
```

Comment:
This looks fragile. Is this really how the spec for styles in KML operates?

Reply:
Resolved in Teams: a StyleMap points to another Style element specified in the KML document. The Placemark.styleUrl can either point to the StyleMap or a Style element.